libs/math/include/math/one_hot.hpp

```diff
@@ -0,0 +1,109 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/assert.hpp"
+
+#include <cassert>
+
+namespace fetch {
+namespace math {
+
+/**
+ * One hot function
+ * @tparam ArrayType
+ * @param ret
+ * @param indices
+ * @param depth
+ * @param on_value
+ * @param off_value
+ * @param axis
+ */
+template <typename ArrayType>
+void OneHot(ArrayType &ret, ArrayType const &indices, typename ArrayType::SizeType depth,
+            typename ArrayType::Type on_value  = typename ArrayType::Type{1.0},
+            typename ArrayType::Type off_value = typename ArrayType::Type{0.0}, SizeType axis = 0)
+{
+  auto ind_it = indices.begin();
+  auto ret_it = ret.Slice().begin();
+
+  if (axis != 0)
+  {
+    // Move the axis we want one-hot to the front
+    // to make it iterable in the inner most loop.
+    ret_it.MoveAxisToFront(axis);
+  }
+
+  while (ind_it.is_valid())
```

Comment:
please add a comment to explain how this while loop works. does it preallocate off_value everywhere and then overwrite some elements? or does it just fill each element correctly first time?